Alloy/Encoders/MaxEncoder.swift

```diff
@@ -0,0 +1,58 @@
+//
+//  MaxEncoder.swift
+//  Adjuster
+//
+//  Created by Eugene Bokhan on 13/02/2019.
+//
+
+import Metal
+import simd
+
+public class MaxEncoder {
+
+    enum Errors: Error {
+        case functionCreationFailed
+        case encoderCreationFailed
+        case invalidBufferStorageMode
+    }
+
+    let pipelineState: MTLComputePipelineState
+
+    public init(library: MTLLibrary) throws {
+        guard
+            let function = library.makeFunction(name: MaxEncoder.functionName)
+        else { throw Errors.functionCreationFailed }
+        let pipelineState = try library.device.makeComputePipelineState(function: function)
+        self.pipelineState = pipelineState
+    }
+
+    public func encode(inputTexture: MTLTexture,
+                       blockSize: BlockSize,
+                       resultBuffer: MTLBuffer,
+                       in commandBuffer: MTLCommandBuffer) throws {
+        guard resultBuffer.storageMode == .shared
+        else { throw Errors.invalidBufferStorageMode }
+
+        commandBuffer.compute { (encoder) in
+            encoder.setTexture(inputTexture, index: 0)
+            encoder.set(blockSize, at: 0)
+            encoder.setBuffer(resultBuffer,
+                              offset: 0,
+                              index: 1)
+
+            let threadsPerThreadgroup = self.pipelineState.max2dThreadgroupSize
+
+            // We have to dispatch only one threadgroup so all threads can share a memory
+            let threadgroupsPerGrid = MTLSize(width: 1, height: 1, depth: 1)
+
+            encoder.setThreadgroupMemoryLength(threadsPerThreadgroup.width * threadsPerThreadgroup.height * 4 * MemoryLayout<Float16>.stride,
+                                               index: 0)
+
+            encoder.setComputePipelineState(self.pipelineState)
+            encoder.dispatchThreadgroups(threadgroupsPerGrid,
```

Comment:
this is not an alloy syntax

Reply:
Yep, Alloy has got a convenience function dispatch2d for that. But due to the logic of this function we have to calculate the covering size argument of this function in a specific way to make the var `count` be equal 1. I think it's better to leave it the it is now.

Reply:
alloy supports providing threadgroup count